fairlearn/preprocessing/_correlation_remover.py

```diff
@@ -0,0 +1,85 @@
+# Copyright (c) Vincent D. Warmerdam, Microsoft Corporation and contributors.
+# Licensed under the MIT License.
+
+import numpy as np
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.utils import check_array
+from sklearn.utils.validation import check_is_fitted
+
+
+class CorrelationRemover(BaseEstimator, TransformerMixin):
+    r"""
+     A component that filters out sensitive correlations in a dataset.
+
+     CorrelationRemover applies a linear transformation to the non-sensitive feature columns in order
+     to remove their correlation with the sensitive feature columns while retaining as much information
+     as possible (as measured by the least-squares error).
+
+     Parameters
+     ----------
+     sensitive_feature_ids : list of columns to filter out this can be a sequence of either int
+       ,in the case of numpy, or string, in the case of pandas.
+     alpha : parameter to control how much to filter, for alpha=1.0 we filter out
+       all information while for alpha=0.0 we don't apply any.
+     center : setting to tell if this preprocessing step should center the data for
+       numerical stability
+
+     Notes
+     -----
+
+     This method will change the original dataset by removing all correlation with sensitive values.
+     To describe that mathematically, let's assume in the original dataset :math:`X` we've got a set of
+     sensitive atttributes :math:`S` and a set of non-sensitive attributes :math:`Z`. Mathmatically this method
+     will be solving the following problem.
+
+     .. math::
+
+        \min _{\mathbf{z}_{1}, \ldots, \mathbf{z}_{n}} \sum_{i=1}^{n}\left\|\mathbf{z}_{i}-\mathbf{x}_{i}\right\|^{2} \\
+        \text{subject to} \\
+        \frac{1}{n} \sum_{i=1}^{n} \mathbf{z}_{i}\left(\mathbf{s}_{i}-\overline{\mathbf{s}}\right)^{T}=\mathbf{0}
+
+
+     The solution to this problem is found by centering sensitive features, fitting a
+     linear regression model to the non-sensitive features and reporting the residual.
+
+     The columns in :math:`S` will be dropped but the hyper parameter :math:`\alpha` does allow you to tweak
+     the amount of filtering that gets applied.
+
+     .. math::
+
+       X_{\text{tfm}} = \alpha X_{\text{filtered}} + (1-\alpha) X_{\text{orig}}
+    """
+
+    def __init__(self, sensitive_feature_ids, alpha=1, center=True):
+        self.sensitive_feature_ids = sensitive_feature_ids
+        self.alpha = alpha
+        self.center = center
+
+    def _split_X(self, X):
+        """Split up X into a sensitive and non-sensitive group."""
+        sensitive = self.sensitive_feature_ids
+        non_sensitive = [i for i in range(X.shape[1]) if i not in sensitive]
+        return X[:, non_sensitive], X[:, sensitive]
+
+    def fit(self, X, y=None):
```

Comment:
The `y=None` is indeed for consistency. Just to confirm though, the `.fit()` method inside fairlearn pipelines send `sensitive_features` as well?

Reply:
`fit` in Fairlearn has `sensitive_features`, but I'm not 100% what you mean by Fairlearn pipelines.

Reply:
Ah right. This might be a larger issue then. 

If we want the tools here to be scikit-learn compatible then this preprocessor needs to be able to the be used in a `Pipeline`. For example, a standard scaler might be applied via; 

```python
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

pipe = Pipeline([
  ('prep', StandardScaler()), 
  ('model', LogisticRegression())
]) 

# The scikit-learn pipeline here does not have `sensitive_features`. 
pipe.fit(X, y) 
```

If we want `CorrelationRemover` to be scikit-learn compatible with regards to the `Pipeline` then we cannot assume to have `sensitive_features` available in the `.fit()`.

Reply:
This goes back to a larger discussion on this topic that @adrinjalali started some time back. If I remember correctly there will be some way to do this with scikit-learn (similar to how you can pass sample_weight), but I'm afraid I'm lacking the context @adrinjalali has.

Reply:
You can pass arbitrary parameters to `fit` in a `Pipeline` via `fit_params`: https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html#sklearn.pipeline.Pipeline.fit

So in this case, it' be like `pipeline.fit(X, y, correlationremover__sensitive_features=blahblah)`

Reply:
In the context of this PR, we should just go with the original proposal, which is:
* provide something like `sensitive_feature_ids` / `sensitive_feature_columns` (pandas terminology?) to the constructor and then we can work with `fit(X,y)` and `predict(X)`

A seeming alternative would be:
* do it similarly to our post-processing step where we have `fit(X,y,sensitive_features=sf)` and also `predict(X,sensitive_features=sf)`.

But this is not compatible with `sklearn.Pipeline`, because I think it only allows `predict_params` at the [very last stage of the pipeline](https://github.com/scikit-learn/scikit-learn/blob/0fb307bf3/sklearn/pipeline.py#L381) (whereas we need them at the beginning).

Reply:
I see, it's true that we're not accepting extra params during `predict` at the moment, and it needs to be fixed. We can't pass those in the constructor because the data passed during `predict` can be very different than the one passed during `fit`. And in reality, we will not have access to the `predict` data during `fit`, therefore we can't pass that to the constructor (think a model deployed  to do predictions in real-time).

Reply:
So I can keep it at `sensitive_feature_ids` as constructor input? Presumably in such a way that it is compatible for both DataFrame and numpy?

Reply:
Yes! Please keep `sensitive_feature_ids` as the constructor argument--with either a list of strings or a list of ints.